### PR TITLE
Fix this reference for functions

### LINKED ISSUE #38720: Custom popover no longer works after upgrading to 5.3.0
### Prerequisites

- [X] I have [searched](https://github.com/twbs/bootstrap/issues?utf8=%E2%9C%93&q=is%3Aissue) for duplicate or closed issues
- [X] I have [validated](https://html5.validator.nu/) any HTML to avoid common problems
- [X] I have read the [contributing guidelines](https://github.com/twbs/bootstrap/blob/main/.github/CONTRIBUTING.md)

### Describe the issue

I used to pass a function to the `content` option of `Popover` as follows, 

```javascript
new bootstrap.Popover(popoverTriggerEl, {
    content: get_popover_content,
  });
```

to trigger a popover defined as 

```html
<a role="button" data-bs-html="true"
        data-bs-trigger="focus" data-bs-container="body" data-bs-toggle="popover"
        data-bs-placement="bottom"
        data-tag-info='{}' >click me</a>
```

where `this.dataset` contains the `data` of the field

```javascript
function get_popover_content() {
  if (this.dataset.tagInfo) {
     //
  }
  //
}
```

This no longer works after upgrading to bootstrap 5.3.0 (`this` points to `document` and `this.dataset` is `null` now) so I am wondering what is the proper way to do it with the new version. Note that the [documentation](https://getbootstrap.com/docs/5.3/components/popovers/) says

> If a function is given, it will be called with its this reference set to the element that the popover is attached to.

and `this` is supposed to be the `<a>` element.

### Reduced test cases

[codepen](https://codepen.io/bopeng/pen/XWyWRra)

The key to look is `console.log(this)`, which points to the document and not the `<a>` element. If you change the `bootstrap` version to `5.2.3`, `this` points to the button itself.
 
### What operating system(s) are you seeing the problem on?

macOS

### What browser(s) are you seeing the problem on?

Chrome

### What version of Bootstrap are you using?

v5.3.0

### PR DESCRIPTION
**Pull request by nwalters512**
### Description

This PR fixes the bug described in #38720. Functions like `content` and `title` weren't being called with the correct `this` reference.

### Motivation & Context

As far as I can tell, #36652 introduced this regression. It switched from calling the function with a specific `this` value to calling it with that value as the first argument.

While this appears to have been unintentional, folks may have started relying on that first argument as a workaround, and thus it could be considered part of the API. I didn't do anything about that yet, but it might be reasonable to retain both the old and new behavior (that is, for functions like `content`, call it with both a specific `this` value _and_ an explicit argument). I can make that change and update the documentation if requested.

I repurposed the existing `execute()` function to support this. If a second argument is provided, it will be an array, and the first value will be treated as the `this` argument. This is done by just passing the entire arguments array to `call`. This required adding `undefined` as the first value in some places that don't expect an specific `this` value.

### Type of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would change existing functionality)

### Checklist

- [x] I have read the [contributing guidelines](https://github.com/twbs/bootstrap/blob/main/.github/CONTRIBUTING.md)
- [x] My code follows the code style of the project _(using `npm run lint`)_
- [x] My change introduces changes to the documentation
- [x] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

### Related issues

Fixes #38720.

### CODE REVIEW

File: js/src/util/template-factory.js [hunk 1]
```diff
@@ -143,7 +143,7 @@ class TemplateFactory extends Config {
   }
 
   _resolvePossibleFunction(arg) {
-    return execute(arg, [this])
+    return execute(arg, [undefined, this])
```
Comment (GeoSot) [inline js/src/util/template-factory.js:146]:
```suggestion
    return typeof arg === 'function' ? arg.call(this._element) : arg
```

I suppose it is better to change only this, as it is count as a regression, and add a test, instead of messing all the rest calls

Comment (nwalters512) [reply js/src/util/template-factory.js:146]: `TemplateFactory` doesn't have access to the underlying element, AFAICT. There are two calls where the `this` argument actually matters:

- `_resolvePossibleFunction` in `js/src/tooltip.js`
- The call to `execute(this._config.placement, ...)` in `_createPopper` in `js/src/tooltip.js`

I can update those two call sites to use the `typeof arg === 'function' ...` pattern if you'd prefer that I avoid changing `execute`?

**Comment by nwalters512** [comment]
@mdo @GeoSot 👋 anything I can do to help this along?

**Comment by nwalters512** [comment]
The failing test appears unrelated to my changes and passes locally for me - is this a flaky test? I'd appreciate it if a maintainer could rerun the workflow.

**Comment by julien-deramond** [comment]
Yes, _JS Tests_ sometimes fail randomly. It's now green again :)

**Comment by mdo** [comment]
Seems good to go @GeoSot?

**Comment by nwalters512** [comment]
@mdo @GeoSot checking in to see if there's anything else I can do to help get this merged!

**Comment by Precidata** [comment]
Happy 1 year anniversary for this bug / PR :)

**Comment by nwalters512** [comment]
@mdo @GeoSot checking in once more! I'd really like to see this land. Let me know if there's anything at all I can do to help this along.

**Comment by nwalters512** [comment]
cc also @julien-deramond and @XhmikosR, looks like you two are the most active maintainers at the moment.

**Comment by julien-deramond** [comment]
Sorry for the very long delay; I know it's frustrating. This PR is on my to-do list along with many other issues and PRs for v5.3.x. Unfortunately, my spare time dedicated to Bootstrap isn't as extensive as I'd like it to be, which is frustrating for me too. Thanks for your patience and your contribution 🙏

**Comment by XhmikosR** [comment]
This needs @GeoSot approval since it was his patch that introduced the breakage. Last time I talked to @GeoSot, he didn't consider this an issue or something like that.

I hate breaking changes myself, but at this point, I'm not sure what the rest of us can do, especially without a test case.

**Comment by nwalters512** [comment]
A breaking change in a minor release is painful, but would have been much less so if the change were documented in the release notes and the documentation for popovers/tooltips/etc. were updated. As it stands, I killed about a day last year trying to figure out why Bootstrap 5.3 broke all my popovers. One way or another, I'd like to save others from that pain.

If you're unwilling to accept this patch, let me know and I'll open a PR with a) tests for the new, current behavior (no `this` reference, instances passed as first argument), b) retroactively updated release notes for 5.3, and c) updated documentation.

**Comment by adriweb** [comment]
If not in a minor release, when would the fix be landing?
Because this also broke prod sites for me, and I've thus sadly had to stick to an older version.

File: js/tests/unit/util/index.spec.js [hunk 2]
```diff
@@ -521,10 +521,10 @@ describe('Util', () => {
 
     it('should execute if arg is function & return the result', () => {
       const functionFoo = (num1, num2 = 10) => num1 + num2
-      const resultFoo = Util.execute(functionFoo, [4, 5])
+      const resultFoo = Util.execute(functionFoo, [undefined, 4, 5])
```
Comment (julien-deramond) [inline js/tests/unit/util/index.spec.js:524]: I'm in the process of understanding what was there before and what's changing in this PR. It's a bit challenging to think about all the use cases, so I hope you won't mind if my remarks or questions are not yet precise or accurate.

If we don't think about the entire context of this PR and focus on this `Util.execute` function only, I'm not that comfortable with an extra first parameter that seems useless. Considering this function in isolation, we would probably expect it to work like this:
* `Util.execute(<function reference>)` or `Util.execute(<function reference>, [])` when there's no args
* `Util.execute(<function reference>, [<arg 1>]`
* `Util.execute(<function reference>, [<arg 1>, <arg 2>]`
* etc.

I'm not saying it doesn't have impacts, or is feasible technically, but in terms of pure interface don't you think this `undefined` first argument in the list should be removed?

Comment (nwalters512) [reply js/tests/unit/util/index.spec.js:524]: I think my rationale for doing it this way is that the `args` array directly mirrors how `Function.prototype.call` works, where the first argument is `this` and the remaining ones are the arguments.

I'm entirely happy to change the interface, e.g. by updating `Util.execute` to explicitly take a value to be used at `this`. For instance, we could update `Util.execute` to look like this:

```ts
const execute = (possibleCallback, thisArg = undefined, args = [], defaultValue = possibleCallback)
```

Then, the test case you commented on here would look like this:

```ts
const resultFoo = Util.execute(functionFoo, undefined, [4, 5]);
```

Of course, we could reorder the arguments, but I think putting it before the arguments (again to mirror `Function.prototype.call`) would make the most sense.

It's also possible that `Util.execute` is simply the wrong thing to be using in the callbacks for tooltips. As mentioned in https://github.com/twbs/bootstrap/pull/38725#discussion_r1255005202, I'm happy to switch those callsites to essentially inline it instead:

```js
typeof arg === 'function' ? arg.call(this._element) : arg

// If we want to retain backwards compatibility with the new API accidentally
// introduced in 5.3, we can do this instead:
typeof arg === 'function' ? arg.call(this._element, this._element) : arg
```

Comment (julien-deramond) [reply js/tests/unit/util/index.spec.js:524]: It wasn't immediately clear to me, but now I see the value in mirroring `Function.prototype.call` as you suggested.

> `const resultFoo = Util.execute(functionFoo, undefined, [4, 5]);`

With your explanation, I'd say that your version in this PR, `Util.execute(functionFoo, [undefined, 4, 5]);` is a better fit because it aligns more closely with `Function.prototype.call`.

> [...] but I think putting it before the arguments (again to mirror Function.prototype.call) would make the most sense.

Yes it would.

> It's also possible that Util.execute is simply the wrong thing to be using in the callbacks for tooltips. As mentioned in https://github.com/twbs/bootstrap/pull/38725#discussion_r1255005202, I'm happy to switch those callsites to essentially inline it instead

I'm not the sole decision-maker here, but from my perspective as a non-JS expert, either approach could work. The initial idea with https://github.com/twbs/bootstrap/pull/36652 was to consolidate functionality and avoid repetitive `typeof arg === 'function'` checks. Thus, retaining `execute` seems reasonable if we want to maintain that approach.

You've already done the work to keep `execute`, so given that I don't have a strong preference at the moment, I'll continue to review the use cases from this angle.

_Note: I'm currently trying to reach the JS maintainers to discuss how we can proceed and finalize this PR with their input and yours._

File: js/src/dropdown.js [hunk 3]
```diff
@@ -320,7 +320,7 @@ class Dropdown extends BaseComponent {
 
     return {
       ...defaultBsPopperConfig,
-      ...execute(this._config.popperConfig, [defaultBsPopperConfig])
+      ...execute(this._config.popperConfig, [undefined, defaultBsPopperConfig])
```
Comment (julien-deramond) [inline js/src/dropdown.js:323]: If we don't apply this modification along with the one in `js/src/tooltip.js`, the unit tests still pass.

```diff
diff --git a/js/src/dropdown.js b/js/src/dropdown.js
index 96094a3e6..9190b3ed5 100644
--- a/js/src/dropdown.js
+++ b/js/src/dropdown.js
@@ -320,7 +320,7 @@ class Dropdown extends BaseComponent {
 
     return {
       ...defaultBsPopperConfig,
-      ...execute(this._config.popperConfig, [undefined, defaultBsPopperConfig])
+      ...execute(this._config.popperConfig, [defaultBsPopperConfig])
     }
   }
 
diff --git a/js/src/tooltip.js b/js/src/tooltip.js
index 24ae2a111..6a6cfeff2 100644
--- a/js/src/tooltip.js
+++ b/js/src/tooltip.js
@@ -438,7 +438,7 @@ class Tooltip extends BaseComponent {
 
     return {
       ...defaultBsPopperConfig,
-      ...execute(this._config.popperConfig, [undefined, defaultBsPopperConfig])
+      ...execute(this._config.popperConfig, [defaultBsPopperConfig])
     }
   }
```

Clearly, these modifications are necessary when calling `execute`. However, this might be an ideal opportunity to add more unit tests to cover these specific use cases. Ensuring comprehensive test coverage now will likely save us time in the future by covering potential edge cases.

Does that make sense? I can help if everyone thinks it's useful.

File: js/tests/unit/util/index.spec.js (hunk 2 above)
Comment (nwalters512) [reply js/tests/unit/util/index.spec.js:524]: Sounds great! I'll be ready to chat with you or the other JS maintainers to find an approach that'll make everyone happy.

Just so it isn't overlooked, a main decision to make here is whether or not Bootstrap will continue calling these functions with the tooltip/popover/etc. as an explicit argument _as well as_ the old behavior of setting `this`. If there's a desire to keep the accidentally-introduced new behavior, I'll add tests and update docs to reflect that.

File: js/src/dropdown.js (hunk 3 above)
Comment (nwalters512) [reply js/src/dropdown.js:323]: Yeah, I'm 100% in favor of adding unit tests for `popperConfig`. I'm happy to let you add them, or I can add them myself since it sounds like you have a lot going on as a maintainer. Just let me know.

Comment (julien-deramond) [reply js/src/dropdown.js:323]: Yeah, I feel like these unit tests will help us remember what's possible and, hopefully, avoid introducing regressions in the future. The approach here would be to initially cover only these two use cases, so we don't go too far in this PR. The common idea is that the new tests pass with this PR, and if we copy them into the _main_ branch, the tests in the _main_ branch will fail.

_Other JS maintainers, if you're reading this, please let us know if we're taking the wrong approach 🙏 Without any feedback, I'll proceed with the review based on my current understanding 😬_

> I'm happy to let you add them, or I can add them myself since it sounds like you have a lot going on as a maintainer. Just let me know.

TBH, if you have some time, I would really appreciate it. But don't feel pressured—if you can't, I'll find time later to take care of it :)

Comment (nwalters512) [reply js/src/dropdown.js:323]: > The common idea is that the new tests pass with this PR, and if we copy them into the main branch, the tests in the main branch will fail.

This is exactly the sort of TDD-way I like to write tests: write them so they fail, update code, watch the tests pass. I'll give that a go here!

Comment (nwalters512) [reply js/src/dropdown.js:323]: Tests added in 03075be. I confirmed they failed if I remove the diff you shared at the start of this thread.

File: js/tests/unit/util/index.spec.js (hunk 2 above)
Comment (julien-deramond) [reply js/tests/unit/util/index.spec.js:524]: OK, so we got a regression where folks had to use a workaround. The basic principle of a fix is that when there's a new patch release (v5.3.x), folks can then remove their workaround so that it works like before.

Since we didn't release a patch promptly, this workaround might now be seen as a de facto part of the API. Even if it isn't officially recognized as such, we may decide not to inconvenience users who have relied on and adapted to this workaround for an extended period. Otherwise, they might perceive the new patch as a disruptive change/breaking change as well, if not retro-compatible. 🙃

Considering these factors, I'd say supporting both the old and the new behavior would ensure a seamless transition for everyone installing the patch version (and we might even consider later on deprecating the old behavior in a future major Bootstrap version, for example):
- Someone migrating from v5.1 to this new v5.3.x could use everything as described in the documentation
- Someone who had to use a workaround could install the new v5.3.x transparently

Yet, I'm wondering whether introducing the new behavior in the documentation should wait until v5.4.0. Even though the feature could be currently present but hidden, delaying its documentation could help us avoid unnecessary remarks and comments. Not entirely sure if this is the right strategy.

Comment (nwalters512) [reply js/tests/unit/util/index.spec.js:524]: Maybe a reasonable middle ground is to implement+test both but avoid documenting the "new" behavior until either a future minor/major release?

Comment (julien-deramond) [reply js/tests/unit/util/index.spec.js:524]: Yeah, that was the idea I had in mind.

1. Implement + test both so that it works for all use cases (migrating from versions before the breaking change, but also supporting transparently workarounds that have been applied). It would give us some space if there are mistakes with the new implementation (as long as the old implementation is fixed and the previous doc still works).

2. Once we have ensured stability, we can document and present the "new" behavior as a minor feature in v5.4.

If this approach sounds reasonable and feasible to you and others involved in this PR, I suggest we proceed with this strategy in mind.

Comment (nwalters512) [reply js/tests/unit/util/index.spec.js:524]: I updated the implementation and tests in a1a57a0.

**Review by julien-deramond — APPROVED** [review]
Thanks a lot again @nwalters512 for this PR and your patience!

Based on our discussions and my review, it looks good to me. I've created [this CodePen](https://codepen.io/julien-deramond/pen/ExBjNqr) to demonstrate that the original issue is resolved. Additionally, I haven't found any regressions during my manual testing, although it's challenging to cover all use cases.

In my opinion, the safest approach to move forward would be to include this in v5.3.4 (along with https://github.com/twbs/bootstrap/pull/40623) and continue our work targeting v5.3.5. If any issues are identified with this fix in the meantime, we can address them in a subsequent v5.3.x release.

I let the @twbs/js-review team finalize this review and/or merge.

**Comment by julien-deramond** [comment]
This topic has been discussed outside of GitHub. In summary, we have mixed feelings because this was an undocumented functionality that some users rely on due to the permissive nature of the interface, leading to an unintended regression when we refactored some aspects. On the other hand, within our JavaScript philosophy, providing internal access to a component is not considered good practice.

Given these factors, the most favorable course of action is to merge this PR for version 5, as it will address the regression for those relying on it. However, we may reconsider or modify this approach in version 6. This could involve changing the JavaScript interface or better documenting these types of usages.

**Comment by nwalters512** [comment]
@julien-deramond @XhmikosR it's been almost a year since the last Bootstrap release. Do you have any estimate for when the next version with this fix will be released?

**Comment by julien-deramond** [comment]
@nwalters512 I'm also looking forward to the release. IMO, we could ship v5.3.4 as is and address other fixes in future v5.3.x updates. That said, I'm not the release manager, so I can't say for sure when it will be released.

**Comment by nwalters512** [comment]
I'm guessing that the answer is no, but as an outsider, is there anything I can do to help push things along?

**Comment by julien-deramond** [comment]
Unfortunately, I don't think so.
